doc/glossary: host, build and target systems

roberth opened the pull request:
… and their relation to Nix

We don't rely on this terminology a lot, but when we do, we should refer to them through here because we need to add our own context.

This will be useful for our hacking documentation and our documentation for [cross compiling nix] (TBD), but I haven't made any references to these terms yet, because another PR is already touching the text.

[cross compiling nix]: https://github.com/NixOS/nix/issues/8561

# Motivation
<!-- Briefly explain what the change is about and why it is desirable. -->

# Context

 - relates to https://github.com/NixOS/nix/pull/8315 but I don't think we need to refer to it directly

<!-- Provide context. Reference open issues if available. -->

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

# Checklist for maintainers

<!-- Contributors: please leave this as is -->

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] documentation in the internal API docs
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

# Priorities

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

roberth committed:
doc/glossary: Briefly describe host, build and target systems and their relation to Nix

We don't rely on this terminology a lot, but when we do, we should
refer to them through here because we need to add our own context.

This will be useful for our hacking documentation and our documentation
for [cross compiling nix] (TBD), but I haven't made any references to
these terms yet, because another PR is already touching the text.

[cross compiling nix]: https://github.com/NixOS/nix/issues/8561

Ericson2314 commented on doc/manual/src/glossary.md:
```diff
@@ -208,6 +208,44 @@
+
+    This is based on the [GCC system terminology](#gloss-gcc-system-terminology).
+
+    Programs that are produced by a [derivation] were configured to have a [build system] that can be executed by the [derivation system](language/derivations.md#attr-system). When using the program, only its [host system] is still relevant.
```
```suggestion
    Programs that are produced by a [derivation] were configured to have a [build system] that can be executed by the [derivation system](language/derivations.md#attr-system).

    Conversely, when using the program after it is been built, the [build system] is no longer relevant.
```
Both host and target (if it exists) are relevant after building.

Ericson2314 commented on doc/manual/src/glossary.md:
```diff
@@ -208,6 +208,44 @@
+
+    A representation of the processor architecture and platform that a compiler will produce, typically a dash-separated string.
+
+    When discussing a build that does not produce a compiler, we only use [build system] and [host system] to describe it; we try to avoid discussing the actual build from the perspective its compiler's configuration.
```
```suggestion
    When discussing a build that does not produce a compiler, we only use [build system] and [host system] to describe it.

    We try to avoid discussing the actual build from the perspective its compiler's configuration, and instead always talk about the build from the perspective of the build itself.
    Avoiding that ensures that if we are building a compiler do use "[target system]", it is clear we are referring to the target of the compiler being built, and not the target of a preexisting compiler used during the build.
```
I don't know if that is too much a digression for Nix, but I do like to make the ambiguity we're avoiding explicit, rather than left to the reader's imagination :).

roberth replied:
```suggestion
    When discussing a build that does not produce a compiler, we only use [build system] and [host system] to describe it.
    
    We try to avoid discussing the actual build from the perspective of its compiler, instead focusing on the build itself.
    Otherwise you won't have a term to distinguish the actual [target system] from what you should have called the [host system].
```

I think I've managed to simplify it. Not sure if I should have tried this in the morning :sweat_smile:

roberth replied:
@Ericson2314 wdyt?

Ericson2314 reviewed (commented):
Thanks for writing this! Just a few minor things.

roberth committed:
doc/glossary: Clarifications in build system

"configured" by itself might be vague if you don't know Autotools and think this could mean configuring the built program at run time.

Both host and target (if it exists) are relevant after building.

Co-authored-by: John Ericson <git@JohnEricson.me>

roberth commented on doc/manual/src/glossary.md:
```diff
@@ -208,6 +208,45 @@
+
+    This is based on the [GCC system terminology](#gloss-gcc-system-terminology).
+
+    Programs that are produced by a [derivation] were configured when they were being built to have a [build system] that can be executed by the [derivation system](language/derivations.md#attr-system).
+    Conversely, when using the program after it is been built, the [build system] is no longer relevant.
```
```suggestion
    A program that is produced by a [derivation] is built with a [build system] that can be executed by the [derivation system](language/derivations.md#attr-system).
    After it has been built, the [build system] is no longer relevant.
```

Simplifying a bit and I don't think it was a particularly important "converse" relationship.

roberth committed:
doc/glossary: Simplify build system text

- Having a build configuration can be implied.
- I don't think this was a particularly important "converse" relationship.

edolstra commented:
This seems more appropriate for the Nixpkgs manual. Nix doesn't really have a concept of host/build/target systems, or indeed of cross-compilation.

Ericson2314 commented:
@edolstra Well one can cross compile Nix itself, so it does have host vs build, just not target. I think @roberth is about to write some docs explaining thing. IMO having a few general things in the glossary doesn't hurt.

fricklerhandwerk reviewed (changes requested):
I think this needs more work to avoid being confusing

roberth committed:
doc/glossary: Define platform and related terms, system -> platform

This is based on the review and discussion earlier today.

Regarding the system / platform confusion we have decided that
"system", the unfortunate GNU term only refers to the string which
partly identifies the more platonic concept of platform.

I find it unfortunate to have to define more terms, but they are
necessary in order to precisely define the Nix terms that we need.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+  - [system type]:{#gloss-system-type}
 
-  - [host system]{#gloss-host-system}
+    An identifier for a specific [platform] or (in rare cases) a class of compatible platforms, such as `x86_64-v3-linux`.
```
```suggestion
    An identifier for a specific [platform] or (in rare cases) a class of compatible platforms, such as `x86_64-linux`.
```

This seems to be atypical for Nix. Why not something everyone knows? Wouldn't be as surprising.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
-    This is based on the [GCC system terminology](#gloss-gcc-system-terminology).
+    Nix is not intimately aware of all possible platforms.
+    Rather, it abstracts the platform into
+     - [system type]{#gloss-system-type} string
```
The system type definition lacks an anchor

```suggestion
     - [system type]string
```

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+    [build platform]: #gloss-build-platform
 
-    Programs that run natively in a [derivation] must be configured with a [host system] that can by executed by the [derivation system](language/derivations.md#attr-system).
+    The [platform] that performs the build or compilation, typically a dash-separated string.
```
```suggestion
    The [platform] that performs the build or compilation, typically denoted by a [system type] string.
```

Same for the other occurrences.

roberth replied:
```suggestion
    The [platform] that performs the build or compilation, typically identified by a [system type] string.
```

Denote would suggest that there's a syntax to it.

**denotation** (computer science) Any mathematical object which describes the meanings of expressions from the languages, formalized in the theory of denotational semantics.

While the system string does have somewhat of a language in practice, it's something Nix does not need to know about, as mentioned elsewhere.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+    The [platform] that performs the build or compilation, typically a dash-separated string.
 
-  - [target system]{#gloss-target-system}
+    A program that is produced by a [derivation] has a [build configuration] with a [build platform] that can be executed by the [derivation system](language/derivations.md#attr-system).
+    After it has been built, the [build platform] is no longer relevant.
```
No need to link the term to itself. All the links are already a strain to read.

As discussed today:

```suggestion
    A program that is produced by a [derivation] must have a [build configuration] where the build platform matches the [`system` field of that derivation](language/derivations.md#attr-system).
    After it has been built, the build platform is no longer relevant.
```

Same for the other occurrences.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+    The [platform] where build products will run on, typically a dash-separated string.
 
-    If the output of a compiler is to be run in a [derivation], the compiler's [target system] must be executable by the [derivation system](language/derivations.md#attr-system).
+    Programs that run natively in a [derivation] must have a [build configuration] with a [host platform] that can by executed by the [derivation system](language/derivations.md#attr-system).
```
```suggestion
    Programs that run when a [derivation] is built must have <as before>
```

Not sure here. Apparently you're using "in a derivation" with the meaning "while running the `builder` executable of a derivation`, which is often abbreviated by "when building a derivation", and I don't think it's semantically wrong, but given that *derivation* is such a mystical term for beginners, we should be careful with condensing too much. I'd rather cut down on "building" a derivation or at least define what that *actually* means in the glossary as well. Actually this is something that I wanted to do some time ago but got rabbit-holed with.

Ericson2314 replied:
I think while the build platform is only a property of the build configuration, the host and target platforms are properties of the built program itself too. I think it makes sense to break symmetry and stop talking about the build configuration for these latter too to further emphasize what information is relevant when.

roberth commented:
> This seems more appropriate for the Nixpkgs manual. Nix doesn't really have a concept of host/build/target systems, or indeed of cross-compilation.

John [has explained well](https://github.com/NixOS/nix/pull/8570#issuecomment-1604185727) why we need these terms, but I'd like to add a couple of points.

We can't just refer to the GNU docs, because they are confusing and they lack context.

Similarly, the Nixpkgs cross compilation docs are highly detailed to a degree that is not appropriate for the Nix manual, or indeed for someone who just wants to use cross compilation.

At least the term _build platform_ is highly related to the _derivation system_.

Clarifying these terms is helpful to anyone who has to deal with them. If Nixpkgs wants to have it, they can, and I'll be glad that we took the thorough effort to get all the details right in a concise, yet complete and precise text.

These terms are relevant to Nix users even if they don't use Nixpkgs. They describe relationships between Nix and the general software ecosystem. The opinionated bits (which are few) may have been discovered through Nixpkgs, but they are good general advice. In other words, these terms are not specific to Nixpkgs.

Finally these definitions are not subject to change. They are thoroughly embedded into the wider ecosystem, and we anticipate no change. Also, work on these terms needs to be done anyway as part of the ecosystem wide documentation efforts, so this is neither a maintenance problem nor a one time resource allocation problem.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+    We try to avoid discussing the actual build from the perspective of its compiler, instead focusing on the build itself.
+    Otherwise we won't have a term to distinguish the actual [target platform] from what we should have called the [host platform].
+
+    If the output of a compiler is to be run in a [derivation], the compiler's [target platform] must be executable by the [derivation system](language/derivations.md#attr-system).
```
I think here the "match" terminology would be especially helpful.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+
+    [target platform]: #gloss-target-platform
+
+    The [platform] for which a compiler will generate code, typically a dash-separated string.
+
+    When discussing a build that does not produce a compiler, we only use [build platform] and [host platform] to describe it. This also applies to compilers with a dynamic target, which is not part of the [build configuration].
```
This should be rephrased, highlighting that it only applies to a few special cases where the target platform is baked into the compiler binary, such as GCC (link to where it matters).

The way it's currently phrased implies it's the most natural thing, and somewhat unusual to have a "dynamic target". Which I think is absolutely not intuitive if you haven't been digging in the bowels of obscure C code in the past 40 years.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+
+    When discussing a build that does not produce a compiler, we only use [build platform] and [host platform] to describe it. This also applies to compilers with a dynamic target, which is not part of the [build configuration].
+
+    We try to avoid discussing the actual build from the perspective of its compiler, instead focusing on the build itself.
+    Otherwise we won't have a term to distinguish the actual [target platform] from what we should have called the [host platform].
```
```suggestion
```

I don't understand what this is trying to tell me.

Ericson2314 replied:
A lot of compiler writers screwed this up, because they only thought of the terms from *their* perspective and not *their user's* perspective.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
     The Nix ecosystem uses the [GCC system terminology](https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html) for the names and roles of _build_, _host_, and _target_, but may apply the term _cross_ more loosely.
 
-    See [host system], [build system] and [target system] for how each relates to [derivation].
+    See [host platform], [build platform] and [target platform] for how each relates to [derivation].
```
```suggestion
    See [host platform], [build platform] and [target platform] for how each relates to a [derivation].
```

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+
+    In the context of Nix, the [build configuration] can be thought of as a [derivation] minus the source code, plus intentional build impurities.
+
+  - [build product]{#gloss-build-product}
```
```suggestion
  - [build result]{#gloss-build-product}
```

We started calling it that in the architecture docs with @Ericson2314 and it's already a term in the manual and nix.dev. It doesn't meen it have to be, but I like it because it doesn't raise mathematical associations.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+    The software configuration whose effects enable the build process and most of which also influence the [build product](#gloss-build-product).
+
+    Build configuration does not have a single agreed upon definition, but consists of a potentially broad set of settings that include:
+    - [build platform], often implicitly
```
```suggestion
    - [build platform], often implicitly meaning the current system
```

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+    Build configuration does not have a single agreed upon definition, but consists of a potentially broad set of settings that include:
+    - [build platform], often implicitly
+    - [host platform]
+    - [target platform], only when building a compiler with a static target platform
```
```suggestion
    - [target platform], only for compilers where their compilation target platform is determined when the compiler itself is build
```

This is a bit lengthy and clumsy, but  think we should make that unambiguous. Maybe you have a more elegant way of saying this without obscuring the meaning.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+    - dependency options, such as which build tools and dependencies to use
+    - anything specified in a custom build script
+
+    In the context of Nix, the [build configuration] can be thought of as a [derivation] minus the source code, plus intentional build impurities.
```
```suggestion
```

This seems misleading. Which source code? In reality, the build configuration is part of the source code. I'd rather leave that out.

We want to help build simplified mental models that don't have to be revised substantially as one becomes more expert.

Ericson2314 replied:
The build configuration is a context for *something*. I am having a hard time finding a reference for this (https://ncatlab.org/nlab/show/context lol) but contexts are sort of dual to expressions --- not the nice self-contained thing, but the expansive thing with no clear external boundary but a nice clear slot in the middle.

It sounds to me what robert is saying "take the thing we are trying to build" out of the donut hole, now what are we left with? That sounds like a good intuition for me. Even if the Nix code is first-party, not 3rd party packaging like Nixpkgs, we can stil distinguish the parts of hte source code that evaluate to derivations from the parts that get copied to the store and referenced from within derivations.

fricklerhandwerk commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+
+  - [build product]{#gloss-build-product}
+
+    The result of running a build process. When a build process runs in a [derivation], build products are typically copied into the [output path](#gloss-output-path)(s).
```
```suggestion
    The result of running a build process. When a `builder` executable of a [derivation] is run, build products are typically copied into the [output path](#gloss-output-path)(s).
```

I don't like this yet though. It ties into defining what "building" means. We use that term very liberally and just assume people know what it means precisely. But it has a precise definition that's just not stated anywhere.

fricklerhandwerk reviewed (changes requested):
I agree with @roberth on the premise. Nixpkgs can figure out and document the mechanics of *facilitating* cross compilation, but the general concepts are deeply embedded into how derivations work (maybe except the target platform, but it's a reality we have to deal with *somewhere*).

The gist of this is really good now. Game-changer-good. I'd give it another round of polishing after these suggestions though.

Ericson2314 commented on doc/manual/src/glossary.md:
```diff
@@ -208,44 +208,85 @@
+    [platform]: #gloss-platform
 
-    A representation of the processor architecture and platform that performs the build or compilation, typically a dash-separated string.
+    A set of execution environments capable of executing some set of software. A platform is typically described by attributes such as instruction set architecture, operating system, and application binary interface.
```
"What does "a set of" mean?
```suggestion
    An execution environments capable of executing some set of software. A platform is typically described by attributes such as instruction set architecture, operating system, and application binary interface.
```
It reads fine to me without it

fricklerhandwerk replied:
I suppose this is meant to cover emulating environments as well. I think it's correct but maybe too clever.

Ericson2314 replied:
Hmm I still wouldn't use set then. I don't think there is an *objective* basis with which to say e.g. "linux and windows syscalls via winelib is 2 interfaces, linux syscalls is one interface". They both are interfaces, it is just interfaces form partial orders.

(Sets are a way to get partial orders out of discrete/unordered things, but I am saying we never had discrete/unordered things to begin with. The concept of sets is not needed.)

tomberek commented:
stale: re-open and re-base as needed
